src/parquet/util/schema-util.h

```diff
@@ -0,0 +1,84 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef PARQUET_SCHEMA_UTIL_H
+#define PARQUET_SCHEMA_UTIL_H
+
+#include <string>
+#include <vector>
+#include <unordered_set>
+
+#include "parquet/exception.h"
+#include "parquet/schema.h"
+#include "parquet/types.h"
+#include "parquet/util/logging.h"
+
+using parquet::ParquetException;
+using parquet::SchemaDescriptor;
+using parquet::schema::GroupNode;
+using parquet::schema::NodePtr;
+using parquet::schema::Node;
+using parquet::LogicalType;
+
+inline bool str_endswith_tuple(const std::string& str) {
+  if (str.size() >= 6) { return str.substr(str.size() - 6, 6) == "_tuple"; }
+  return false;
+}
+
+// Special case mentioned in the format spec:
+//   If the name is array or ends in _tuple, this should be a list of struct
+//   even for single child elements.
+inline bool HasStructListName(const GroupNode* node) {
+  return (node->name() == "array" ||
+          str_endswith_tuple(node->name()));
+}
+
+// TODO(itaiin): This aux. function is to be deleted once repeated structs are supported
+inline bool IsSimpleStruct(const NodePtr& node) {
+  if (!node->is_group()) return false;
+  if (node->is_repeated()) return false;
+  if (node->logical_type() == LogicalType::LIST) return false;
+  // Special case mentioned in the format spec:
+    //   If the name is array or ends in _tuple, this should be a list of struct
+    //   even for single child elements.
+  GroupNode* group = static_cast<GroupNode*>(node.get());
+  if (group->field_count() == 1 && HasStructListName(group)) return false;
+
+  return true;
+}
+
+// Coalesce a list of schema fields indices which are the roots of the
+// columns referred by a list of column indices
+inline bool ColumnIndicesToFieldIndices(const SchemaDescriptor& descr,
```

Comment:
Thanks for the clarifying. 

> Our intent in the Arrow API was to only reference these semantic column numbers

This intent is unclear to me as SchemaDescriptor only uses leaves to get Column, see code below:
```cpp
const ColumnDescriptor* SchemaDescriptor::Column(int i) const {
  DCHECK(i >= 0 && i < static_cast<int>(leaves_.size()));
  return &leaves_[i];
}
```
I used to think column reference to leaf node until I saw your clarification and some parquet-mr code digging.

Back to column vs field indices, I believe it's clear what field means in arrow::Schema, the top(base) node, which would be {foo, foo2, foo3, foo4} in your example, and I think it's much easier that if there is a ReadField api so that we can simply read corresponding field by index.

Also for API usage, it would be quite common to selected nodes(field or column) of interested by user. For example, user only wants to read the first and third field in the file(foo and foo3 in your example), or by name path: {foo, foo2, foo4.qux}, {foo.bar, foo4}.  To be able to support that,  I think we should also do column index mapping.
> if you were to read {0, 1, 2} then the reader must deal with reading the tree for the 3 leaf nodes.

I think we should merge this read request to generate an arrow array constrained by `Schema<Field foo>`